pkg/tar/tar.go

```diff
@@ -165,7 +165,10 @@ func extractFile(tr *tar.Reader, hdr *tar.Header, overwrite bool) error {
 		return fmt.Errorf("unsupported type: %v", typ)
 	}
 
-	if err := os.Lchown(p, hdr.Uid, hdr.Gid); err != nil {
+	if uidCount != 0 && (uint64(hdr.Uid) >= uidCount || uint64(hdr.Gid) >= uidCount) {
+		return fmt.Errorf("archive contains out of range uid/gid")
+	}
+	if err := os.Lchown(p, hdr.Uid+int(uidShift), hdr.Gid+int(uidShift)); err != nil {
```

Comment:
That would be another argument for factoring out those checks to UidRange member function.